wait_for/__init__.py

```diff
@@ -117,6 +126,9 @@ def wait_for(func, func_args=[], func_kwargs={}, logger=None, **kwargs):
     Raises:
         TimedOutError: If num_sec is exceeded after an unsuccessful func() invocation.
     """
+    # Hide this call in the detailed traceback
```

Comment:
Hiding the call to wait_for? Is this a good idea?

Reply:
Yes. it hides the lengthy frame from wait_for that actually gives no info (as wait for generaly won't rise an exception unless it timeouts). Look how much more readable it is with
```
================================================ FAILURES ================================================
_____________________________________________ test_wait_for ______________________________________________

    def test_wait_for():
>       wait_for.wait_for(method, num_sec=.1, handle_exception=True)
E       wait_for.TimedOutError: Could not do function method() at /home/jhenner/work/wait_for/test_wait_for.py:3 in time

test_wait_for.py:7: TimedOutError
======================================== 1 failed in 1.06 seconds ========================================
```
than of without it:
```
================================================ FAILURES ================================================
_____________________________________________ test_wait_for ______________________________________________

    def test_wait_for():
>       wait_for.wait_for(method, num_sec=.1, handle_exception=True)

test_wait_for.py:7: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

func = <function method at 0x7f7d8143f9d8>, func_args = [], func_kwargs = {}
logger = <Logger wait_for.default (WARNING)>, kwargs = {'handle_exception': True, 'num_sec': 0.1}
st_time = 1554475855.2366219, total_time = 1, num_sec = 0.1, expo = False
f_code = <code object method at 0x7f7d81415d20, file "/home/jhenner/work/wait_for/test_wait_for.py", line 3>
line_no = 3, filename = '/home/jhenner/work/wait_for/test_wait_for.py'

    def wait_for(func, func_args=[], func_kwargs={}, logger=None, **kwargs):
        """Waits for a certain amount of time for an action to complete
        Designed to wait for a certain length of time,
        either linearly in 1 second steps, or exponentially, up to a maximum.
        Returns the output from the function once it completes successfully,
        along with the time taken to complete the command.
        Note: If using the expo keyword, the returned elapsed time will be inaccurate
            as wait_for does not know the exact time that the function returned
            correctly, only that it returned correctly at last check.
        Args:
            func: A function to be run
            func_args: A list of function arguments to be passed to func
            func_kwargs: A dict of function keyword arguments to be passed to func
            num_sec: An int describing the number of seconds to wait before timing out.
            timeout: Either an int describing the number of seconds to wait before timing out. Or a
                :py:class:`timedelta` object. Or a string formatted like ``1h 10m 5s``. This then sets
                the ``num_sec`` variable.
            expo: A boolean flag toggling exponential delay growth.
            message: A string containing a description of func's operation. If None,
                defaults to the function's name.
            fail_condition: An object describing the failure condition that should be tested
                against the output of func. If func() == fail_condition, wait_for continues
                to wait. Can be a callable which takes the result and returns boolean whether to fail.
                You can also specify it as a  set, that way it checks whether it is present in the
                iterable.
            handle_exception: A boolean controlling the handling of excepetions during func()
                invocation. If set to True, in cases where func() results in an exception,
                clobber the exception and treat it as a fail_condition.
            delay: An integer describing the number of seconds to delay before trying func()
                again.
            fail_func: A function to be run after every unsuccessful attempt to run func()
            quiet: Do not write time report to the log (default False)
            very_quiet: Do not log unless there was an error (default False). Implies quiet.
            silent_failure: Even if the entire attempt times out, don't throw a exception.
        Returns:
            A tuple containing the output from func() and a float detailing the total wait time.
        Raises:
            TimedOutError: If num_sec is exceeded after an unsuccessful func() invocation.
        """
        # Hide this call in the detailed traceback
        # https://docs.pytest.org/en/latest/example/simple.html#writing-well-integrated-assertion-helpers
        #__tracebackhide__ = True
        logger = logger or default_hidden_logger
        st_time = time.time()
        total_time = 0
    
        num_sec = _get_timeout_secs(kwargs)
        expo = kwargs.get('expo', False)
    
        f_code, line_no, filename, message = _get_context(func, kwargs.get('message', None))
    
        fail_condition = kwargs.get('fail_condition', False)
        fail_condition_check = _get_failcondition_check(fail_condition)
        handle_exception = kwargs.get('handle_exception', False)
        delay = kwargs.get('delay', 1)
        fail_func = kwargs.get('fail_func', None)
        very_quiet = kwargs.get("very_quiet", False)
        quiet = kwargs.get("quiet", False) or very_quiet
        silent_fail = kwargs.get("silent_failure", False)
    
        t_delta = 0
        tries = 0
        if not very_quiet:
            logger.debug('Started {} at {}'.format(message, st_time))
        while t_delta <= num_sec:
            try:
                tries += 1
                out = func(*func_args, **func_kwargs)
            except Exception as e:
                logger.info("wait_for hit an exception: {}: {}".format(type(e).__name__, e))
                if handle_exception:
                    out = fail_condition
                    logger.info("Call failed with following exception, but continuing "
                                "as handle_exception is set to True")
                else:
                    logger.info(
                        "Wait for {} took {} tries and {} seconds "
                        "before failure from an exception.".format(
                            message, tries, time.time() - st_time))
                    raise
            if out is fail_condition or fail_condition_check(out):
                time.sleep(delay)
                total_time += delay
                if expo:
                    delay *= 2
                if fail_func:
                    fail_func()
            else:
                duration = time.time() - st_time
                if not quiet:
                    logger.debug('Took {:0.2f} to do {}'.format(duration, message))
                if not very_quiet:
                    logger.debug(
                        'Finished {} at {}, {} tries'.format(message, st_time + t_delta, tries))
                return WaitForResult(out, duration)
            t_delta = time.time() - st_time
        if not very_quiet:
            logger.debug('Finished at {}'.format(st_time + t_delta))
        if not silent_fail:
            logger.error("Couldn't complete {} at {}:{} in time, took {:0.2f}, {} tries".format(message,
                filename, line_no, t_delta, tries))
            logger.error('The last result of the call was: {}'.format(out))
>           raise TimedOutError("Could not do {} at {}:{} in time".format(message, filename, line_no))
E           wait_for.TimedOutError: Could not do function method() at /home/jhenner/work/wait_for/test_wait_for.py:3 in time
```

Reply:
Note that only the stack fram from wait_for is hidden. See an example when handle_exception=False. The frames from function the wait_for is executing are shown.
```
================================================ FAILURES ================================================
_____________________________________________ test_wait_for ______________________________________________

    def test_wait_for():
>       wait_for.wait_for(method, num_sec=.1, handle_exception=False)

test_wait_for.py:7: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def method():
>       raise Exception
E       Exception

test_wait_for.py:4: Exception
======================================== 1 failed in 0.05 seconds ========================================
(py33) ➜  wait_for git:(display_lambda) ✗ 
```